### PR TITLE
Add asymmetric key support for MailServer requests

### PR DESCRIPTION
**Pull request by adambabik**
In a swarm `#280-disc-mail-servers`, mail servers will be discoverable using a discovery protocol. Each mail server will be verified against a smart contract. In such a setup, it does not make sense to use a password as that password would need to be put in the smart contract as well or made public in some other way, thus a support for the asymmetric key is added.

Requesting client will use MailServer's public key by default.

The support for the symmetric is left intact for backward compatibility.

### CODE REVIEW

File: mailserver/mailserver.go [hunk 1]
```diff
@@ -136,19 +138,29 @@ func (s *WMailServer) setupLimiter(limit time.Duration) {
...
+// setupRequestMessageDecryptor setup a Whisper filter to decrypt
+// incoming Whisper requests.
+func (s *WMailServer) setupRequestMessageDecryptor(config *params.WhisperConfig) error {
+	var filter whisper.Filter
 
-	s.key, err = s.w.GetSymKey(MailServerKeyID)
-	if err != nil {
-		return fmt.Errorf("save symmetric key: %s", err)
+	if config.Password != "" {
+		keyID, err := s.w.AddSymKeyFromPassword(config.Password)
```
Comment (adriacidre) [inline mailserver/mailserver.go]: maybe I'll add this logic on its own method `symkey, err := s.symKeyFromPassword(config.Password)`

File: node/node.go [hunk 2]
```diff
@@ -223,12 +223,18 @@ func activateShhService(stack *node.Node, config *params.NodeConfig, db *leveldb
 
 		// enable mail service
 		if config.WhisperConfig.EnableMailServer {
-			if config.WhisperConfig.Password == "" {
+			if config.WhisperConfig.Password == "" && config.WhisperConfig.PasswordFile != "" {
```
Comment (adriacidre) [inline node/node.go]: PasswordFile emptiness is already being checked inside [ReadPasswordFile](https://github.com/status-im/status-go/pull/1075/files#diff-a894bbf771e5541f902b5b16f367f99dL126), so, this is probably not needed.

**Review by adriacidre — APPROVED** [review]
Nice! left some small comments.

I guess the plan is to progressively get rid of password encryption in favour of asym key based encryption, should we add a deprecation comment to discourage its usage?

File: params/config.go [hunk 3]
```diff
@@ -92,6 +95,12 @@ type WhisperConfig struct {
 	// (if no account file selected, then this password is used for symmetric encryption).
 	Password string
 
+	// AsymKeyFile is a file with an asymmetric key to decrypt messages sent to MailServer.
```
Comment (adriacidre) [inline params/config.go]: If they're mail server related config options shouldn't we prefix them with MailServer like [MailServerRateLimit ](https://github.com/status-im/status-go/blob/2b4a7dfb1c56b02783a884c43af98677361dc3fc/params/config.go#L118) or [MailServerCleanupPeriod](https://github.com/status-im/status-go/blob/2b4a7dfb1c56b02783a884c43af98677361dc3fc/params/config.go#L121)?

File: mailserver/mailserver_test.go [hunk 4]
```diff
@@ -61,67 +63,143 @@ func (s *MailserverSuite) SetupTest() {
...
-			info:          "Initializing a mail server with a config with empty DataDir and inactive limiter",
+			info:          "config with both asym key and password",
+		},
+		{
+			config: params.WhisperConfig{
+				DataDir:             s.config.DataDir,
+				MailServerRateLimit: 5,
+			},
+			expectedError: nil,
+			info:          "config witih rate limit",
```
Comment (gravityblast) [inline mailserver/mailserver_test.go]: just a small typo in `witih`

**Comment by gravityblast** [comment]
@adambabik  I noticed running the tests it fails with a panic

**Comment by adambabik** [comment]
@pilu @adriacidre I fixed all issues reported in the comments.

**Comment by adriacidre** [comment]
All looking good :)

**Review by gravityblast — APPROVED** [review]
LGTM!

**Pull request merged**